qemu/tests/numa_opts.py

```diff
@@ -1,49 +1,120 @@
+        """
+        The numa info in guest os, linux only
+
+        return: An array of (ram, cpus) tuples, where ram is the RAM size in
+                MB and cpus is a set of CPU numbers
+        """
+
+        numa_info_guest = NumaInfo(session=session)
+
+        numa_guest = []
+        nodes_guest = numa_info_guest.online_nodes
+        for node in nodes_guest:
+            node_size = numa_info_guest.online_nodes_meminfo[node]['MemTotal']
+            node_size = float(normalize_data_size('%s KB' % node_size))
+            node_cpus = numa_info_guest.online_nodes_cpus[node]
+            node_cpus = set([int(v) for v in node_cpus.split()])
+            numa_guest.append((node_size, node_cpus))
+
+        # It is a known WONTFIX issue for x86, node info of node0 and node1 is
+        # opposite in guest os when vm have 2 nodes
+        if (vm_arch in ("x86_64", "i686") and len(numa_guest) == 2):
+            numa_guest.reverse()
+        return numa_guest
 
     vm = env.get_vm(params["main_vm"])
+    os_type = params["os_type"]
+    vm_arch = params["vm_arch_name"]
+    session = vm.wait_for_login()
 
+    error_context.context("starting numa_opts test...", logging.info)
+
+    # Get numa info from monitor with "info monitor"
     numa = vm.monitors[0].info_numa()
-    dbg("info numa reply: %r", numa)
+    error_context.context("info numa reply: %r" % numa, logging.info)
+
+    if os_type == 'linux':
+        if not utils_package.package_install('numactl', session):
+            test.cancel("Please install numactl to proceed")
+        # Get numa info in guest os with "numactl -H",  only for Linux
+        numa_guest = numa_info_guest()
+        error_context.context("numactl -H reply: %r" % numa_guest, logging.info)
+
+        # Use 30 plus the gap of 'MemTotal' in OS and '-m' in cli as threshold
+        MemTotal = get_mem_info(session, 'MemTotal')
+        MemTotal = float(normalize_data_size('%s KB' % MemTotal))
+        error_context.context("MemTotal in guest os is %s MB"
+                              % MemTotal, logging.info)
+        threshold = float(params.get_numeric("mem") - MemTotal) + 30
+        error_context.context("The acceptable threshold is: %s"
+                              % threshold, logging.info)
+        session.close()
+    else:
+        numa_guest = numa
 
     numa_nodes = params.get("numa_nodes")
     if numa_nodes:
-        numa_nodes = int(params.get("numa_nodes"))
+        numa_nodes = int(numa_nodes)
         if len(numa) != numa_nodes:
-            test.fail(
-                "Wrong number of numa nodes: %d. Expected: %d" %
-                (len(numa), numa_nodes))
+            test.fail("[Monitor]Wrong number of numa nodes: %d. Expected: %d" %
+                      (len(numa), numa_nodes))
 
-    for nodenr, node in enumerate(numa):
+    if os_type == 'linux':
+        guest_expect_nodes = params.get("guest_expect_nodes", numa_nodes)
+        guest_expect_nodes = int(guest_expect_nodes)
```

Comment:
Above 2 lines could combine to one?
`guest_expect_nodes = int(params.get("guest_expect_nodes", numa_nodes))`